### PR TITLE
Changes for VB-3878 - check if there are any higher priority legal or medical appointments.

### PR DESCRIPTION
**Pull request by dhirajshettymoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerService.kt
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerService.kt:1]: moved all Sessions related methods to new service - VisitSchedulerSessionService - no other change on this file.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/controller/OrchestrationSessionsController.kt [hunk 1]
```diff
@@ -119,8 +119,13 @@ class OrchestrationSessionsController(private val visitSchedulerService: VisitSc
     )
     @NullableNotEmpty(message = "An empty visitors list is not allowed")
     visitors: List<Long>? = null,
+    @RequestParam(value = "withAppointmentsCheck", required = false)
+    @Parameter(
+      description = "Defaults to true if not passed. If true, will not return visit times that clash with higher priority legal or medical appointments.",
+    )
+    withAppointmentsCheck: Boolean? = true,
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/controller/OrchestrationSessionsController.kt:126]: have kept this non mandatory for now as am not sure if PVB needs this flag. Will check with Simon / Omar.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/dto/whereabouts/enums/HigherPriorityMedicalOrLegalEvents.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  MEDICAL_OTHER("MEOT", "Medical - Other"),
+  MEDICAL_PSYCHOLOGY_SERVICES("MEPS", "Medical_Psychology Services"),
+  MEDICAL_X_RAY("MEXR", "Medical - X-ray"),
+  PRE_RELEASE_MEETING("PRM", "Pre-Release Meeting"),
+  PSYCHOLOGY_SERVICES_FORENSIC("FOPS", "Psychology Services - Forensic"),
+  SOLICITOR_MEETINGS_NON_LEGAL_VISIT("SOLS", "Solicitor Meetings (Not Legal Visit)"),
+  VIDEO_LINK_COURT_HEARING("VLB", "Video Link - Court Hearing"),
+  VIDEO_LINK_LEGAL_APPOINTMENT("VLLA", "Video Link - Legal Appointment"),
+  VIDEO_LINK_PAROLE_HEARING("VLPA", "Video Link - Parole Hearing"),
+  VIDEO_LINK_PROBATION_MEETING("VLPM", "Video Link - Probation Meeting"),
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/dto/whereabouts/enums/HigherPriorityMedicalOrLegalEvents.kt:29]: https://dsdmoj.atlassian.net/wiki/spaces/PSCH/pages/4897734794/Legal+and+Medical+Appointments+Mapping

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    visitors: List<Long>?,
+  ): SessionRestriction {
+    return if (prisonerProfileService.hasPrisonerGotClosedRestrictions(prisonerId) ||
+      (if (visitors != null) prisonerProfileService.hasVisitorsGotClosedRestrictions(prisonerId, visitors) else false)
+    ) {
+      CLOSED
+    } else {
+      requestedSessionRestriction ?: OPEN
+    }
+  }
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt:113]: moved from VisitSchedulerService.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    return events.joinToString(" and ")
+  }
+
+  private fun hasOverlappingTimes(
+    scheduledEventStartTime: LocalTime,
+    scheduledEventEndTime: LocalTime,
+    visitSessionStartTime: LocalTime,
+    visitSessionEndTime: LocalTime,
+  ): Boolean {
+    return (
```
Comment (adaviesMOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt:158]: return (
    visitSessionStartTime < scheduledEventEndTime &&
    visitSessionEndTime > scheduledEventStartTime
);

Could this be simplified to these conditions? As there is a bit of overlap with the 3 conditionals. If we have these 2 checks, it should cover it all. Double check with the existing unit tests

Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt:158]: also might be possible to make it more readable by doing

return (
visitSessionStartTime.isBefore(scheduledEventEndTime) && visitSessionEndTime.isAfter(scheduledEventStartTime)
);

**Review by adaviesMOJ — APPROVED** [review]
LGTM, only 1 suggestion. Amazing work mate!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt (hunk 4 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt:158]: nice. the tests work. need to get my head around how it works first :)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/dto/whereabouts/enums/HigherPriorityMedicalOrLegalEvents.kt (hunk 2 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/dto/whereabouts/enums/HigherPriorityMedicalOrLegalEvents.kt:29]: Yikes load of types there

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    val LOG: Logger = LoggerFactory.getLogger(this::class.java)
+    const val CLASHING_APPOINTMENT_LOG_MSG =
+      "Visit session for prisonerId - {}, date - {}, start time - {}, end time - {} is unavailable as it clashes with {} medical / legal appointment(s), appointment details - {}"
+  }
+
+  fun getVisitSessions(prisonCode: String, prisonerId: String?, min: Int?, max: Int?): List<VisitSessionDto>? {
+    return visitSchedulerClient.getVisitSessions(prisonCode, prisonerId, min, max)
+  }
+
+  fun getAvailableVisitSessions(
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt:36]: This method is getting a bit long

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        visitors?.let { prisonerProfileService.getBannedRestrictionDateRage(prisonerId, visitors, dateRange) }
+          ?: dateRange
+      visitSchedulerClient.getAvailableVisitSessions(prisonCode, prisonerId, sessionRestriction, updatedDateRange)
+    } catch (e: DateRangeNotFoundException) {
+      LOG.error("getAvailableVisitSessions range is not returned therefore we do not have a valid date range and should return an empty list")
+      emptyList()
+    }
+
+    availableVisitSessions = if (withAppointmentsCheck && availableVisitSessions.isNotEmpty()) {
+      val higherPriorityAppointments = appointmentsService.getHigherPriorityAppointments(prisonerId, dateRange.fromDate, dateRange.toDate)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt:57]: any chance you could lift this out and put it in a sub method

**Review by AledWEvans — COMMENTED** [review]
Some comments, minor changes

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt (hunk 4 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt:146]: Updated. Thanks!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt (hunk 5 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt:37]: Agreed, have updated and moved code to smaller methods.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt (hunk 6 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt]: Updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/controller/OrchestrationSessionsController.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/controller/OrchestrationSessionsController.kt:126]: Checked with Simon who said he is happy with this approach

**Review by AledWEvans — APPROVED** [review]
LGTM

**Pull request merged**